plugins/catalog-backend-module-incremental-ingestion/src/database/IncrementalIngestionDatabaseManager.ts

```diff
@@ -285,40 +298,37 @@ export class IncrementalIngestionDatabaseManager {
    * @param ingestionId - string
    * @returns All entities to remove for this burst.
    */
-  async computeRemoved(provider: string, ingestionId: string) {
+  async computeRemoved(provider: string) {
+    const previousIngestion = await this.getPreviousIngestionRecord(provider);
     return await this.client.transaction(async tx => {
-      const removed: { entity: string; ref: string }[] = await tx(
-        'final_entities',
-      )
-        .select(
-          tx.ref('final_entity').as('entity'),
-          tx.ref('refresh_state.entity_ref').as('ref'),
-        )
-        .join(
-          'refresh_state',
-          'refresh_state.entity_id',
-          'final_entities.entity_id',
-        )
-        .join('search', 'search.entity_id', 'final_entities.entity_id')
-        .whereNotIn(
-          'entity_ref',
-          tx('ingestion_marks')
-            .join(
-              'ingestion_mark_entities',
-              'ingestion_marks.id',
-              'ingestion_mark_entities.ingestion_mark_id',
-            )
-            .select('ingestion_mark_entities.ref')
-            .where('ingestion_marks.ingestion_id', ingestionId),
-        )
-        .andWhere(
-          'search.key',
-          `metadata.annotations.${INCREMENTAL_ENTITY_PROVIDER_ANNOTATION}`,
-        )
-        .andWhere('search.value', provider);
-      return removed.map(entity => {
-        return { entity: JSON.parse(entity.entity) };
-      });
+      if (previousIngestion) {
+        const rows: { ref: string }[] = await tx('ingestion_mark_entities')
+          .select('ingestion_mark_entities.ref')
+          .join(
+            'ingestion_marks',
+            'ingestion_marks.id',
+            'ingestion_mark_entities.ingestion_mark_id',
+          )
+          .join('ingestions', 'ingestions.id', 'ingestion_marks.ingestion_id')
+          .where('ingestions.id', previousIngestion.id);
```

Comment:
As I mentioned elsewhere, I think for the purposes of deletion, we can't consider just _a single ingestion_, so this logic should probably be tweaked.

Correct me if I'm wrong but, as I understand it, what happens during ingestion is that groups of entities (successfully fetched pages) are effectively upserted into the catalog, which is a persistent, permanent modification. And this happens whether some later page starts failing, or we time out, or whatnot. This tells me that we need to always, unconditionally, keep track of the union of _all_ such upserts that we've requested. That is,

For each page read:
- issue a `'delta'` with some items `added`
- If that succeeds, upsert into a global tracking table that holds the union of all refs we've ever issued
- And also add those refs into a local tracking table that holds the union of all refs issued in this particular run

If/when we reach the last page of the current run successfully,
- Compute the difference between the local tracking table and the global tracking table - ie find all refs we have issued at some point ever, and the refs we issued this time
- Issue a `'delta'` with those items `removed`
- If that succeeds, remove those refs from the global tracking table

Reply:
Ah but I didn't mean it _never_ deletes records from the global table. I meant that it contains whatever the module thinks _should_ be in `refresh_state` at this point in time, to the best of its knowledge, caused by itself. I.e. anything it has successfully committed minus everything it has successfully removed. Note my last point "If that succeeds, remove those refs from the global tracking table".

I believe that it's important that that particular table is burst-agnostic, because it represents an externality (the refresh_state table) which doesn't know of what bursts are. Therefore I think that if this idea is correct, it might entail some changes to this module's internal table structure.

Then separately from this burst-agnostic table, indeed there needs to be a burst-aware table that records what happens in the current burst as well. That's the one that ultimately lets us compute the diff between this burst and the externality.

Reply:
Got it. OK, I think we're on the same page then, because I'm envisioning the `ingestion_mark_entities` table as exactly that: A representation of what the provider thinks the `refresh_state` table should have in it. Or more specifically, what the table should have in it that is there due to the actions of the provider.

With that in mind, I'm not sure the table structure needs to change, because all the information we need seems to be there already:

* Every record in the `ingestion_mark_entities` table is associated with a record in the `ingestion_marks` table.
* Every record in the `ingestion_marks` table is in turn associated with a record in the `ingestions` table. These records are associated with a page in the current burst.
* To update which ingestion a record in the `ingestion_mark_entities` table is associated with, just update it with the mark ID for the current page at the time of ingestion.
* If we abide by this flow, the records in the `ingestion_mark_entities` table for a given provider will always represent 1) the current status of entities in the `refresh_state` table for the provider, and 2) any entities that haven't been updated in the current ingestion yet (and will be removed from the `refresh_state` table if the ingestion completes without updating them).

Does that fit the bill?

Reply:
Hm. Yeah my primary concern was that there was a chain of references all the way up to `ingestions`, and that I seemed to remember wanting to clean out ingestions over time. As long as there won't be any cascading deletes that removes references to things we've committed, it is probably fine.